gammapy/cube/psf_map.py

```diff
@@ -0,0 +1,189 @@
+import numpy as np
+import astropy.units as u
+from gammapy.irf import EnergyDependentTablePSF, PSF3D
+from gammapy.maps import WcsGeom, WcsNDMap, MapAxis, Map
+
+__all__ = [
+    'make_psf3d',
+    'make_psf_map',
+    'PSFMap'
+]
+
+def make_psf3d(psf_analytical, rad):
+    """ Creates a PSF3D from an analytical PSF.
+
+    Parameters
+    ----------
+    psf_analytical : `~gammapy.irf.EnergyDependentMultiGaussPSF`
+        the analytical PSF to be transformed to PSF3D
+    rad : `~astropy.unit.Quantity` or `~astropy.coordinates.Angle`
+        the array of position errors (rad) on which the PSF3D will be defined
+
+    Return
+    ------
+    psf3d : `~gammapy.irf.PSF3D`
+        the PSF3D. It will be defined on the same energy and offset values than the input psf.
+    """
+    offsets = psf_analytical.theta
+    energy = psf_analytical.energy
+    energy_lo = psf_analytical.energy_lo
+    energy_hi = psf_analytical.energy_hi
+    rad_lo = rad[:-1]
+    rad_hi = rad[1:]
+
+    psf_values = np.zeros((rad_lo.shape[0],offsets.shape[0],energy_lo.shape[0]))*u.Unit('sr-1')
+
+    for i,offset in enumerate(offsets):
+        psftable = psf_analytical.to_energy_dependent_table_psf(offset)
+        psf_values[:,i,:] = psftable.evaluate(energy,0.5*(rad_lo+rad_hi)).T
+
+    return PSF3D(energy_lo, energy_hi, offsets, rad_lo, rad_hi, psf_values,
+                 psf_analytical.energy_thresh_lo, psf_analytical.energy_thresh_hi)
+
+
+def make_psf_map(psf, pointing, ref_geom, max_offset):
+    """Make a psf map for a single observation
+
+    Expected axes : rad and true energy in this specific order
+
+    Parameters
+    ----------
+    psf : `~gammapy.irf.PSF3D`
+        the PSF IRF
+    pointing : `~astropy.coordinates.SkyCoord`
+        the pointing direction
+    ref_geom : `~gammapy.maps.MapGeom`
+        the map geom to be used. It provides the target geometry.
+        rad and true energy axes should be given in this specific order.
+    max_offset : `~astropy.coordinates.Angle`
+        maximum offset w.r.t. fov center
+
+    Returns
+    -------
+    psfmap : `~gammapy.maps.Map`
+        the resulting PSF map
+    """
+    # retrieve energies. This should be properly tested
+    energy = ref_geom.axes[1].center * ref_geom.axes[1].unit
+
+    # retrieve rad from psf IRF directly (this might be given as an argument instead)
+    rad = ref_geom.axes[0].center * ref_geom.axes[0].unit
+
+    # Compute separations with pointing position
+    separations = pointing.separation(ref_geom.to_image().get_coord().skycoord)
+    valid = np.where(separations < max_offset)
+
+    # Compute PSF values
+    psf_values = np.transpose(psf.evaluate(offset=separations[valid], energy=energy, rad=rad), axes=(2, 0, 1))
+    psfmap = Map.from_geom(ref_geom, unit='sr-1')
+    psfmap.data[:, :, valid[0], valid[1]] += psf_values.to(psfmap.unit).value
+    return psfmap
+
+
+class PSFMap():
```

Comment:
Write `class PSFMap(object)`.

If you omit the `object` you'll get an "old-style class" on Python 2. Usually you don't notice, but really those old-style classes should not be used since Python 2.1 or something like that, and then in some cases they do behave in weird ways.

Reply:
My bad...